File: desktop-app/src/main/java/com/heartguard/desktop/api/InfluxDBService.java
package com.heartguard.desktop.api;

import com.influxdb.client.InfluxDBClient;
import com.influxdb.client.InfluxDBClientFactory;
import com.influxdb.query.FluxRecord;
import com.influxdb.query.FluxTable;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Cliente para consultar datos de series temporales desde InfluxDB.
 * Lee signos vitales de pacientes en tiempo real.
 */
public class InfluxDBService {
    private final String url;
    private final String token;
    private final String org;
    private final String bucket;
    private InfluxDBClient client;

    private static final int CONNECTION_TIMEOUT_SECONDS = 5;
    private static final int READ_TIMEOUT_SECONDS = 10;
    private boolean useMockData = false;
    
    public InfluxDBService(String url, String token, String org, String bucket) {
        this.url = url;
        this.token = token;
        this.org = org;
        this.bucket = bucket;
    }

    /**
     * Conectar al servidor InfluxDB
     */
    public void connect() {
        if (client == null) {
            System.out.println("[InfluxDB] Connecting to: " + url);
            System.out.println("[InfluxDB] Organization: " + org);
            System.out.println("[InfluxDB] Bucket: " + bucket);
            System.out.println("[InfluxDB] Timeout: " + CONNECTION_TIMEOUT_SECONDS + "s");
            
            try {
                // Crear cliente InfluxDB con configuración básica
                client = InfluxDBClientFactory.create(url, token.toCharArray(), org, bucket);
                
                // Verificar conexión con una query simple
                System.out.println("[InfluxDB] Testing connection...");
                try {
                    client.getQueryApi().query("buckets()", org);
                    System.out.println("[InfluxDB] ✓ Connection established and verified successfully");
                    useMockData = false;
                } catch (Exception testEx) {
                    System.err.println("[InfluxDB] ⚠ WARNING: Connection established but cannot query.");
                    System.err.println("[InfluxDB] Error: " + testEx.getMessage());
                    System.err.println("[InfluxDB] Will use mock data for demonstration");
                    useMockData = true;
                }
            } catch (Exception e) {
                System.err.println("[InfluxDB] ✗ ERROR connecting: " + e.getMessage());
                System.err.println("[InfluxDB] Will use mock data for demonstration");
                useMockData = true;
                client = null;
            }
        } else {
            System.out.println("[InfluxDB] Already connected");
        }
    }

    /**
     * Desconectar del servidor InfluxDB
     */
    public void disconnect() {
        if (client != null) {
            client.close();
            client = null;
        }
    }

    /**
     * Obtener los datos de signos vitales de un paciente en las últimas N horas.
     * 
     * @param patientId ID del paciente
     * @param hoursBack Cuántas horas hacia atrás consultar
     * @return Lista de lecturas de signos vitales
     */
    public List<VitalSignsReading> getPatientVitalSigns(String patientId, int hoursBack) {
        List<VitalSignsReading> readings = new ArrayList<>();

        if (client == null) {
            System.out.println("InfluxDB client not connected, connecting now...");
            connect();
        }

        try {
            // Flux query para obtener los datos del paciente
            String flux = String.format("""
                from(bucket: "%s")
                  |> range(start: -%dh)
                  |> filter(fn: (r) => r["_measurement"] == "vital_signs")
                  |> filter(fn: (r) => r["patient_id"] == "%s")
                  |> filter(fn: (r) => 
                      r["_field"] == "heart_rate" or
                      r["_field"] == "spo2" or
                      r["_field"] == "systolic_bp" or
                      r["_field"] == "diastolic_bp" or
                      r["_field"] == "temperature" or
                      r["_field"] == "gps_longitude" or
                      r["_field"] == "gps_latitude"
                  )
                  |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")
                  |> sort(columns: ["_time"], desc: false)
                """, bucket, hoursBack, patientId);

            List<FluxTable> tables = client.getQueryApi().query(flux);

            for (FluxTable table : tables) {
                for (FluxRecord record : table.getRecords()) {
                    VitalSignsReading reading = new VitalSignsReading();
                    reading.timestamp = record.getTime();

                    // Extraer valores de los campos
                    Object heartRate = record.getValueByKey("heart_rate");
                    Object spo2 = record.getValueByKey("spo2");
                    Object systolicBp = record.getValueByKey("systolic_bp");
                    Object diastolicBp = record.getValueByKey("diastolic_bp");
                    Object temperature = record.getValueByKey("temperature");
                    Object gpsLong = record.getValueByKey("gps_longitude");
                    Object gpsLat = record.getValueByKey("gps_latitude");

                    if (heartRate != null) reading.heartRate = ((Number) heartRate).intValue();
                    if (spo2 != null) reading.spo2 = ((Number) spo2).intValue();
                    if (systolicBp != null) reading.systolicBp = ((Number) systolicBp).intValue();
                    if (diastolicBp != null) reading.diastolicBp = ((Number) diastolicBp).intValue();
                    if (temperature != null) reading.temperature = ((Number) temperature).doubleValue();
                    if (gpsLong != null) reading.gpsLongitude = ((Number) gpsLong).doubleValue();
                    if (gpsLat != null) reading.gpsLatitude = ((Number) gpsLat).doubleValue();

                    readings.add(reading);
                }
            }
        } catch (Exception e) {
            System.err.println("Error querying InfluxDB: " + e.getMessage());
            e.printStackTrace();
        }

        return readings;
    }

    /**
     * Obtener los últimos N registros de signos vitales de un paciente.
     * Útil para gráficas en tiempo real con ventana deslizante.
     * 
     * @param patientId ID del paciente
     * @param limit Número de registros a obtener
     * @return Lista de lecturas de signos vitales
     */
    public List<VitalSignsReading> getLatestPatientVitalSigns(String patientId, int limit) {
        List<VitalSignsReading> readings = new ArrayList<>();

        if (client == null) {
            System.out.println("[InfluxDB] Client not connected, connecting now...");
            connect();
        }
        
        // Si debemos usar datos mock, generarlos
        if (useMockData || client == null) {
            System.out.println("[InfluxDB] ⚠ Using mock data (InfluxDB not available)");
            return generateMockData(patientId, limit);
        }
        
        System.out.println("[InfluxDB] Querying for patient: " + patientId + " (last " + limit + " readings)");

        try {
            // Flux query para obtener los últimos N registros
            String flux = String.format("""
                from(bucket: "%s")
                  |> range(start: -24h)
                  |> filter(fn: (r) => r["_measurement"] == "vital_signs")
                  |> filter(fn: (r) => r["patient_id"] == "%s")
                  |> filter(fn: (r) => 
                      r["_field"] == "heart_rate" or
                      r["_field"] == "spo2" or
                      r["_field"] == "systolic_bp" or
                      r["_field"] == "diastolic_bp" or
                      r["_field"] == "temperature" or
                      r["_field"] == "gps_longitude" or
                      r["_field"] == "gps_latitude"
                  )
                  |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")
                  |> sort(columns: ["_time"], desc: true)
                  |> limit(n: %d)
                """, bucket, patientId, limit);

            System.out.println("[InfluxDB] Executing Flux query:");
            System.out.println(flux);
            
            List<FluxTable> tables = client.getQueryApi().query(flux);
            
            System.out.println("[InfluxDB] Query returned " + tables.size() + " tables");

            for (FluxTable table : tables) {
                System.out.println("[InfluxDB] Processing table with " + table.getRecords().size() + " records");
                for (FluxRecord record : table.getRecords()) {
                    VitalSignsReading reading = new VitalSignsReading();
                    reading.timestamp = record.getTime();

                    Object heartRate = record.getValueByKey("heart_rate");
                    Object spo2 = record.getValueByKey("spo2");
                    Object systolicBp = record.getValueByKey("systolic_bp");
                    Object diastolicBp = record.getValueByKey("diastolic_bp");
                    Object temperature = record.getValueByKey("temperature");
                    Object gpsLong = record.getValueByKey("gps_longitude");
                    Object gpsLat = record.getValueByKey("gps_latitude");

                    if (heartRate != null) reading.heartRate = ((Number) heartRate).intValue();
                    if (spo2 != null) reading.spo2 = ((Number) spo2).intValue();
                    if (systolicBp != null) reading.systolicBp = ((Number) systolicBp).intValue();
                    if (diastolicBp != null) reading.diastolicBp = ((Number) diastolicBp).intValue();
                    if (temperature != null) reading.temperature = ((Number) temperature).doubleValue();
                    if (gpsLong != null) reading.gpsLongitude = ((Number) gpsLong).doubleValue();
                    if (gpsLat != null) reading.gpsLatitude = ((Number) gpsLat).doubleValue();

                    readings.add(reading);
                    System.out.println("[InfluxDB] Read: " + reading);
                }
            }

            // Revertir para tener orden cronológico ascendente
            readings.sort((a, b) -> a.timestamp.compareTo(b.timestamp));
            
            System.out.println("[InfluxDB] Successfully retrieved " + readings.size() + " readings for patient " + patientId);
        } catch (Exception e) {
            System.err.println("[InfluxDB] ERROR querying: " + e.getMessage());
            e.printStackTrace();
        }

        return readings;
    }

    /**
     * Generar datos mock para demostración cuando InfluxDB no está disponible
     */
    private List<VitalSignsReading> generateMockData(String patientId, int limit) {
        List<VitalSignsReading> mockReadings = new ArrayList<>();
        Instant now = Instant.now();
        
        System.out.println("[InfluxDB] Generating " + limit + " mock readings for patient " + patientId);
        
        for (int i = limit - 1; i >= 0; i--) {
            VitalSignsReading reading = new VitalSignsReading();
            
            // Timestamp: cada 10 minutos hacia atrás
            reading.timestamp = now.minus(i * 10, java.time.temporal.ChronoUnit.MINUTES);
            
            // Generar valores realistas con pequeñas variaciones
            double variation = Math.sin(i * 0.5) * 5; // Variación sinusoidal
            
            reading.heartRate = (int) (72 + variation + (Math.random() * 6 - 3));
            reading.spo2 = (int) (97 + (Math.random() * 2));
            reading.systolicBp = (int) (120 + variation + (Math.random() * 8 - 4));
            reading.diastolicBp = (int) (80 + variation * 0.5 + (Math.random() * 6 - 3));
            reading.temperature = 36.5 + (Math.random() * 0.6 - 0.3);
            reading.gpsLatitude = 25.6866 + (Math.random() * 0.01 - 0.005);
            reading.gpsLongitude = -100.3161 + (Math.random() * 0.01 - 0.005);
            
            mockReadings.add(reading);
        }
        
        System.out.println("[InfluxDB] Generated " + mockReadings.size() + " mock readings");
        return mockReadings;
    }
    
    /**
     * Clase que representa una lectura de signos vitales
     */
    public static class VitalSignsReading {
        public Instant timestamp;
        public int heartRate;
        public int spo2;
        public int systolicBp;
        public int diastolicBp;
        public double temperature;
        public double gpsLongitude;
        public double gpsLatitude;

        public VitalSignsReading() {
        }

        @Override
        public String toString() {
            return String.format("VitalSigns[time=%s, HR=%d, SpO2=%d, BP=%d/%d, Temp=%.2f°C]",
                    timestamp, heartRate, spo2, systolicBp, diastolicBp, temperature);
        }
    }
}

File: micro-services/user/src/user/services/influxdb_service.py
"""
Cliente de InfluxDB para consultas de datos en tiempo real
"""
from influxdb_client import InfluxDBClient
from influxdb_client.client.query_api import QueryApi
from typing import List, Dict, Any, Optional
from datetime import datetime, timedelta
import logging

from ..config import get_config

logger = logging.getLogger(__name__)


class InfluxDBService:
    """Servicio para interactuar con InfluxDB"""
    
    def __init__(self):
        config = get_config()
        self.url = config.INFLUXDB_URL
        self.token = config.INFLUXDB_TOKEN
        self.org = config.INFLUXDB_ORG
        self.bucket = config.INFLUXDB_BUCKET
        self._client = None
        self._query_api = None
    
    @property
    def client(self) -> InfluxDBClient:
        """Obtiene o crea el cliente de InfluxDB"""
        if self._client is None:
            try:
                self._client = InfluxDBClient(
                    url=self.url,
                    token=self.token,
                    org=self.org
                )
                logger.info(f"✓ Cliente InfluxDB conectado: {self.url}")
            except Exception as e:
                logger.error(f"Error conectando a InfluxDB: {e}")
                raise
        return self._client
    
    @property
    def query_api(self) -> QueryApi:
        """Obtiene la API de consultas"""
        if self._query_api is None:
            self._query_api = self.client.query_api()
        return self._query_api
    
    def get_patient_realtime_data(
        self, 
        patient_id: str, 
        hours: int = 1,
        limit: int = 100
    ) -> List[Dict[str, Any]]:
        """
        Obtiene datos en tiempo real de un paciente
        
        Args:
            patient_id: UUID del paciente
            hours: Ventana de tiempo en horas (default: 1 hora)
            limit: Límite de puntos de datos (default: 100)
            
        Returns:
            Lista de puntos de datos con timestamp y valores
        """
        try:
            # Construir query de Flux
            query = f'''
            from(bucket: "{self.bucket}")
                |> range(start: -{hours}h)
                |> filter(fn: (r) => r["_measurement"] == "vital_signs")
                |> filter(fn: (r) => r["patient_id"] == "{patient_id}")
                |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")
                |> sort(columns: ["_time"], desc: true)
                |> limit(n: {limit})
            '''
            
            logger.debug(f"Ejecutando query InfluxDB para paciente {patient_id}")
            
            # Ejecutar query
            tables = self.query_api.query(query, org=self.org)
            
            # Procesar resultados
            results = []
            for table in tables:
                for record in table.records:
                    data_point = {
                        'timestamp': record.get_time().isoformat(),
                        'gps_longitude': record.values.get('gps_longitude'),
                        'gps_latitude': record.values.get('gps_latitude'),
                        'heart_rate': record.values.get('heart_rate'),
                        'spo2': record.values.get('spo2'),
                        'systolic_bp': record.values.get('systolic_bp'),
                        'diastolic_bp': record.values.get('diastolic_bp'),
                        'body_temperature': record.values.get('temperature'),
                        'heart_rate_alert': int(record.values.get('heart_rate_alert', 0)),
                        'spo2_alert': int(record.values.get('spo2_alert', 0)),
                        'bp_alert': int(record.values.get('bp_alert', 0)),
                        'temperature_alert': int(record.values.get('temperature_alert', 0)),
                    }
                    results.append(data_point)
            
            logger.info(f"✓ Obtenidos {len(results)} puntos de datos para paciente {patient_id}")
            return results
            
        except Exception as e:
            logger.error(f"Error consultando InfluxDB para paciente {patient_id}: {e}")
            return []
    
    def get_patient_latest_vitals(self, patient_id: str) -> Optional[Dict[str, Any]]:
        """
        Obtiene los últimos signos vitales de un paciente
        
        Args:
            patient_id: UUID del paciente
            
        Returns:
            Diccionario con los últimos valores o None
        """
        try:
            query = f'''
            from(bucket: "{self.bucket}")
                |> range(start: -24h)
                |> filter(fn: (r) => r["_measurement"] == "vital_signs")
                |> filter(fn: (r) => r["patient_id"] == "{patient_id}")
                |> pivot(rowKey:["_time"], columnKey: ["_field"], valueColumn: "_value")
                |> sort(columns: ["_time"], desc: true)
                |> limit(n: 1)
            '''
            
            tables = self.query_api.query(query, org=self.org)
            
            for table in tables:
                for record in table.records:
                    return {
                        'timestamp': record.get_time().isoformat(),
                        'gps_longitude': record.values.get('gps_longitude'),
                        'gps_latitude': record.values.get('gps_latitude'),
                        'heart_rate': record.values.get('heart_rate'),
                        'spo2': record.values.get('spo2'),
                        'systolic_bp': record.values.get('systolic_bp'),
                        'diastolic_bp': record.values.get('diastolic_bp'),
                        'body_temperature': record.values.get('temperature'),
                        'heart_rate_alert': int(record.values.get('heart_rate_alert', 0)),
                        'spo2_alert': int(record.values.get('spo2_alert', 0)),
                        'bp_alert': int(record.values.get('bp_alert', 0)),
                        'temperature_alert': int(record.values.get('temperature_alert', 0)),
                    }
            
            return None
            
        except Exception as e:
            logger.error(f"Error obteniendo últimos vitales para paciente {patient_id}: {e}")
            return None
    
    def get_patient_vitals_summary(
        self, 
        patient_id: str, 
        hours: int = 24
    ) -> Dict[str, Any]:
        """
        Obtiene resumen estadístico de signos vitales
        
        Args:
            patient_id: UUID del paciente
            hours: Ventana de tiempo en horas
            
        Returns:
            Diccionario con estadísticas (promedio, min, max)
        """
        try:
            query = f'''
            from(bucket: "{self.bucket}")
                |> range(start: -{hours}h)
                |> filter(fn: (r) => r["_measurement"] == "vital_signs")
                |> filter(fn: (r) => r["patient_id"] == "{patient_id}")
                |> filter(fn: (r) => 
                    r["_field"] == "heart_rate" or 
                    r["_field"] == "spo2" or 
                    r["_field"] == "systolic_bp" or 
                    r["_field"] == "diastolic_bp" or
                    r["_field"] == "temperature"
                )
                |> group(columns: ["_field"])
            '''
            
            # Calcular estadísticas
            mean_query = query + '|> mean()'
            min_query = query + '|> min()'
            max_query = query + '|> max()'
            
            results = {
                'mean': {},
                'min': {},
                'max': {},
                'period_hours': hours
            }
            
            # Ejecutar queries
            for stat_type, stat_query in [('mean', mean_query), ('min', min_query), ('max', max_query)]:
                tables = self.query_api.query(stat_query, org=self.org)
                for table in tables:
                    for record in table.records:
                        field = record.values.get('_field')
                        value = record.values.get('_value')
                        if field and value is not None:
                            results[stat_type][field] = round(float(value), 2)
            
            return results
            
        except Exception as e:
            logger.error(f"Error obteniendo resumen de vitales para paciente {patient_id}: {e}")
            return {
                'mean': {},
                'min': {},
                'max': {},
                'period_hours': hours,
                'error': str(e)
            }
    
    def close(self):
        """Cierra la conexión con InfluxDB"""
        if self._client:
            self._client.close()
            logger.info("✓ Cliente InfluxDB cerrado")


# Instancia global del servicio
_influx_service = None


def get_influx_service() -> InfluxDBService:
    """Obtiene o crea la instancia global del servicio InfluxDB"""
    global _influx_service
    if _influx_service is None:
        _influx_service = InfluxDBService()
    return _influx_service
